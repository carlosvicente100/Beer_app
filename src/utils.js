export const DEFAULT_VALUE = -1

export const formatDate = (value) => {
  const date = value.split('/')
  return {
    month: parseInt(date[0]),
    year: parseInt(date[1])
  }
}

export const formatOrderYears = (yearList) => {
  let newArray = yearList
    .filter((element, index, self) => index === self.findIndex((t) => t.year === element.year))
    .map((element) => {
      return { name: element.year, value: element.year }
    })
    .sort(function (a, b) {
      return a.value - b.value
    })

  return newArray
}

export const filterByDate = (startDate, endDate, items) => {
  return items.filter((item) => {
    const itemDate = formatDate(item.first_brewed)

    const validYear = (itemDate) =>
      (itemDate.year > startDate.year || (itemDate.year === startDate.year && itemDate.month >= startDate.month)) &&
      (itemDate.year < endDate.year || (itemDate.year === endDate.year && itemDate.month <= endDate.month))

    const defaultValues = () =>
      startDate.year === DEFAULT_VALUE &&
      startDate.month === DEFAULT_VALUE &&
      endDate.year === DEFAULT_VALUE &&
      endDate.month === DEFAULT_VALUE
    return validYear(itemDate) || defaultValues()
  })
}

export const makePagination = (itemsFiltered) => {
  return itemsFiltered.reduce((result, value, index) => {
    if (index % 5 !== 0) {
      result[result.length - 1].push(value)
    } else {
      result.push([value])
    }
    return result
  }, [])
}

export const MonthList = [
  { name: 'January', value: 1 },
  { name: 'February', value: 2 },
  { name: 'March', value: 3 },
  { name: 'April', value: 4 },
  { name: 'May', value: 5 },
  { name: 'June', value: 6 },
  { name: 'July', value: 7 },
  { name: 'August', value: 8 },
  { name: 'September', value: 9 },
  { name: 'October', value: 10 },
  { name: 'November', value: 11 },
  { name: 'December', value: 12 }
]
