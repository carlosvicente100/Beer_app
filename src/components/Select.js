import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'

import { DEFAULT_VALUE } from '../utils'

const SelectContainer = styled.div`
  display: flex;
  justify-content: space-between;
  width: 100%;
`
const SelectList = styled.select`
  margin: auto;
`

const Select = ({ values, current, onSelect }) => {
  const selectValue = (event) => {
    const value = parseInt(event.target.value)
    if (value !== current) {
      onSelect(parseInt(value))
    }
  }

  const makeOptionList = values.map((element, index) => (
    <option key={index} value={element.value} defaultValue={element.value === current}>
      {element.name}
    </option>
  ))

  return (
    <SelectContainer className="Select">
      <SelectList data-testid="select" value={current} onChange={selectValue}>
        <option value={DEFAULT_VALUE}>Empty</option>
        {makeOptionList}
      </SelectList>
    </SelectContainer>
  )
}
Select.propTypes = {
  onSelect: PropTypes.func.isRequired,
  values: PropTypes.array.isRequired,
  current: PropTypes.number.isRequired
}

Select.defaultProps = {}

export default Select
