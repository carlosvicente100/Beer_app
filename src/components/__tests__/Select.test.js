import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer'

import 'jest-styled-components'

import Select from '../Select'
import { MonthList, DEFAULT_VALUE } from '../../utils'

const dummyProps = {
  onSelect: jest.fn(),
  values: MonthList,
  current: DEFAULT_VALUE
}

beforeEach(() => {
  jest.clearAllMocks()
})

describe('Select Component Suite', () => {
  test('Snapshot check', () => {
    const component = renderer.create(<Select {...dummyProps} />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('Select Structure', () => {
    test('Should have a default value - empty', () => {
      const SelectComponent = render(<Select {...dummyProps} />)
      expect(SelectComponent.getByText('Empty')).toBeInTheDocument()
    })
  })
  describe('Select Functionallity', () => {
    test('Should call onSelect on change value', () => {
      const SelectComponent = render(<Select {...dummyProps} />)
      // expect(SelectComponent.getByText('Empty')).toBeInTheDocument()
      fireEvent.change(SelectComponent.getByTestId('select'), {
        target: { value: dummyProps.values[4].value }
      })

      expect(dummyProps.onSelect).toHaveBeenCalledTimes(1)
      expect(dummyProps.onSelect).toHaveBeenCalledWith(dummyProps.values[4].value)
    })

    test("Shouldn't call onSelect on select same value", () => {
      const SelectComponent = render(<Select {...dummyProps} current={dummyProps.values[2].value} />)
      expect(dummyProps.onSelect).toHaveBeenCalledTimes(0)

      fireEvent.change(SelectComponent.getByTestId('select'), {
        target: { value: dummyProps.values[2].value }
      })

      expect(dummyProps.onSelect).toHaveBeenCalledTimes(0)
    })
  })
})
