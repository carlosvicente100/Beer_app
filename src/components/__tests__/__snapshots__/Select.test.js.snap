// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component Suite Snapshot check 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.c1 {
  margin: auto;
}

<div
  className="c0 Select"
>
  <select
    className="c1"
    data-testid="select"
    onChange={[Function]}
    value={-1}
  >
    <option
      value={-1}
    >
      Empty
    </option>
    <option
      defaultValue={false}
      value={1}
    >
      January
    </option>
    <option
      defaultValue={false}
      value={2}
    >
      February
    </option>
    <option
      defaultValue={false}
      value={3}
    >
      March
    </option>
    <option
      defaultValue={false}
      value={4}
    >
      April
    </option>
    <option
      defaultValue={false}
      value={5}
    >
      May
    </option>
    <option
      defaultValue={false}
      value={6}
    >
      June
    </option>
    <option
      defaultValue={false}
      value={7}
    >
      July
    </option>
    <option
      defaultValue={false}
      value={8}
    >
      August
    </option>
    <option
      defaultValue={false}
      value={9}
    >
      September
    </option>
    <option
      defaultValue={false}
      value={10}
    >
      October
    </option>
    <option
      defaultValue={false}
      value={11}
    >
      November
    </option>
    <option
      defaultValue={false}
      value={12}
    >
      December
    </option>
  </select>
</div>
`;
